README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis-CI Build
Status](https://travis-ci.org/jjchern/laus.svg?branch=master)](https://travis-ci.org/jjchern/laus)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/jjchern/laus?branch=master&svg=true)](https://ci.appveyor.com/project/jjchern/laus)

    #> Warning: package 'knitr' was built under R version 3.4.4

# About

An R data package contains Local Area Unemployment Statistics (LAUS)
from U.S. Bureau of Labor Statistics (BLS). So far it contains the
following four series:

  - `state_year`:
      - State-level employment status of the civilian noninstitutional
        population, annual average series (1976-2018)
  - `state_month_sa`:
      - State-level employment status of the civilian noninstitutional
        population, seasonaly adjusted monthly series (1976-2019.03)
  - `state_month_nsa`:
      - State-level employment status of the civilian noninstitutional
        population, not seasonaly adjusted monthly series (1976-2019.03)
  - `county_year`:
      - County-level employment status of the civilian noninstitutional
        population, annual average series (1990-2018)
  - `county_month_nsa`:
      - County-level employment status of the civilian noninstitutional
        population, not seasonaly adjusted monthly series (1976-2019.02)

## Relevant packages

  - [`rUnemploymentData`](https://github.com/trulia/rUnemploymentData/)

# Installation

``` r
# install.package("remotes")
remotes::install_github("jjchern/laus@v0.0.4")
```

# Usage

## List all series

``` r
data(package = "laus")$results[,3]
#> [1] "county_month_nsa" "county_year"      "state_month_nsa" 
#> [4] "state_month_sa"   "state_year"
```

## List all variable names and variable labels for `state_year`

``` r
names(laus::state_year)
#>  [1] "fips"      "state"     "year"      "pop"       "clf"      
#>  [6] "pc_clf"    "emp"       "pc_emp"    "unem"      "unem_rate"

# devtools::install_github("larmarange/labelled")
labelled::var_label(laus::state_year)
#> $fips
#> [1] "FIPS code"
#> 
#> $state
#> [1] "State or area"
#> 
#> $year
#> [1] "Year"
#> 
#> $pop
#> [1] "Civilian non-institutional population"
#> 
#> $clf
#> [1] "Total number of people in civilian labor force"
#> 
#> $pc_clf
#> [1] "Labor force participation rate (= labor force / population; Age: 16 years and over)"
#> 
#> $emp
#> [1] "Total number of people employed"
#> 
#> $pc_emp
#> [1] "Employment-population ratio (= employment / population; Age: 16 years and over)"
#> 
#> $unem
#> [1] "Total number of people unemployed"
#> 
#> $unem_rate
#> [1] "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
```

## Show the data frame

``` r
library(tidyverse)
laus::state_year
#> # A tibble: 2,091 x 10
#>    fips  state   year     pop    clf pc_clf    emp pc_emp   unem unem_rate
#>    <chr> <chr>   <chr>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>     <dbl>
#>  1 01    Alabama 1976  2.63e6 1.50e6   57.0 1.40e6   53.1 1.02e5      6.80
#>  2 02    Alaska  1976  2.40e5 1.64e5   68.2 1.51e5   63.0 1.24e4      7.60
#>  3 04    Arizona 1976  1.65e6 9.87e5   59.8 8.91e5   54.0 9.61e4      9.70
#>  4 05    Arkans… 1976  1.55e6 8.92e5   57.7 8.29e5   53.6 6.22e4      7.00
#>  5 06    Califo… 1976  1.58e7 9.90e6   62.6 8.99e6   56.8 9.05e5      9.10
#>  6 08    Colora… 1976  1.86e6 1.24e6   66.8 1.17e6   62.9 7.24e4      5.80
#>  7 09    Connec… 1976  2.26e6 1.46e6   64.4 1.32e6   58.4 1.35e5      9.30
#>  8 10    Delawa… 1976  4.21e5 2.65e5   62.9 2.42e5   57.4 2.32e4      8.80
#>  9 11    Distri… 1976  5.17e5 3.35e5   64.8 3.05e5   59.0 3.02e4      9.00
#> 10 12    Florida 1976  6.51e6 3.60e6   55.3 3.27e6   50.3 3.30e5      9.20
#> # ... with 2,081 more rows
laus::county_year
#> # A tibble: 86,869 x 11
#>    laus_code  fips  state_fips county_fips county  state year  labor_force
#>    <chr>      <chr> <chr>      <chr>       <chr>   <chr> <chr>       <dbl>
#>  1 CN0100100… 01001 01         001         Autaug… AL    1990       16908.
#>  2 CN0100300… 01003 01         003         Baldwi… AL    1990       46824.
#>  3 CN0100500… 01005 01         005         Barbou… AL    1990       11490.
#>  4 CN0100700… 01007 01         007         Bibb C… AL    1990        7434.
#>  5 CN0100900… 01009 01         009         Blount… AL    1990       19168.
#>  6 CN0101100… 01011 01         011         Bulloc… AL    1990        4407.
#>  7 CN0101300… 01013 01         013         Butler… AL    1990        8838.
#>  8 CN0101500… 01015 01         015         Calhou… AL    1990       51519.
#>  9 CN0101700… 01017 01         017         Chambe… AL    1990       17705.
#> 10 CN0101900… 01019 01         019         Cherok… AL    1990        9644.
#> # ... with 86,859 more rows, and 3 more variables: employed <dbl>,
#> #   unemployed <dbl>, unemployment_rate <dbl>
```

## Plot thematic maps with the unemployment data

<!-- # ```{r state-map, echo=FALSE} -->

<!-- # # Combine state map data with unemployment rate -->

<!-- # # devtools::install_github("jjchern/usmapdata") -->

<!-- # usmapdata::state %>%  -->

<!-- #   inner_join(laus::state_year, by = c("id" = "fips")) -> unem -->

<!-- #  -->

<!-- # # Plot a Map -->

<!-- #  -->

<!-- # library(gganimate) -->

<!-- #  -->

<!-- # ggplot() + -->

<!-- #   geom_map(data = unem, map = unem, -->

<!-- #            aes(x = long, y = lat, map_id = id, fill = unem_rate, frame = year), -->

<!-- #            colour = alpha("white", 0.5), size=0.2) + -->

<!-- #   coord_map("albers", lat0 = 30, lat1 = 40) + -->

<!-- #   viridis::scale_fill_viridis(option = "B") + -->

<!-- #   ggtitle("US Unemplyment Rates (%) by State in") + -->

<!-- #   ggthemes::theme_map() + -->

<!-- #   theme(legend.position = c(.85, .3), -->

<!-- #         legend.title=element_blank(),  -->

<!-- #         plot.title = element_text(hjust = 0.5)) -> g -->

<!-- #  -->

<!-- # gganimate(g) -->

<!-- # ``` -->

![state-map](README/README-fig-state-map-.gif)

<!-- # ```{r county-map, echo=FALSE} -->

<!-- # usmapdata::county %>%  -->

<!-- #         inner_join(laus::county_year, by = c("id" = "fips")) -> unem_county -->

<!-- #  -->

<!-- # ggplot() + -->

<!-- #   geom_map(data = unem_county, map = unem_county, -->

<!-- #            aes(x = long, y = lat, map_id = id, fill = unemployment_rate, frame = year), -->

<!-- #            colour = alpha("white", 0.1), size=0.2) + -->

<!-- #   geom_map(data = usmapdata::state, map = usmapdata::state, -->

<!-- #            aes(x = long, y = lat, map_id = id), -->

<!-- #            colour = alpha("white", 0.75), fill = "NA") +         -->

<!-- #   coord_map("albers", lat0 = 30, lat1 = 40) + -->

<!-- #   viridis::scale_fill_viridis(option = "B") + -->

<!-- #   ggtitle("US Unemplyment Rates (%) by County in") + -->

<!-- #   ggthemes::theme_map() + -->

<!-- #   theme(legend.position = c(.85, .3), -->

<!-- #         legend.title=element_blank(),  -->

<!-- #         plot.title = element_text(hjust = 0.5)) -> c -->

<!-- #  -->

<!-- # gganimate(c) -->

<!-- # ``` -->

![county-map](README/README-fig-county-map-.gif)

## Plot selected Series

![](README/README-fig-ts_unem_03_12_mon-1.png)<!-- -->

![](README/README-fig-ts_unem_03_12_ann-1.png)<!-- -->

![](README/README-fig-il-1.png)<!-- -->


NEWS.md

# laus 0.0.4

* Update all series in the package
    - county monthly series (up to 2019-02)
    - county annual series (up to 2018)
    - state annual series (up to 2018)
    - state monthly series (up to 2019-03)
    
# laus 0.0.3

* Update all series in the package
    - county monthly series (up to 2017.10)
    - county annual series (up to 2016)
    - state annual series (up to 2016)
    - state monthly series (up to 2017.11)

# laus 0.0.2

* Added `laus::county_month_nsa`

# laus 0.0.1

* Initial release


R/data.R
#' @title State by year level US employment data
#'
#' @description Annual averages series of employment status of the civilian noninstitutional population from BLS.
#'
#' @format A data frame with ten variables:
#' * `fips` = "FIPS code",
#' * `state` = "State or area",
#' * `year` = "Year",
#' * `pop` = "Civilian non-institutional population",
#' * `clf` = "Total numer of people in civilian labor force",
#' * `pc_clf` = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
#' * `emp` = "Total number of people employed",
#' * `pc_emp` = "Employment-population ratio (= employment / population; Age: 16 years and over)",
#' * `unem` = "Total number of people unemployed",
#' * `unem_rate` = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
#'
#' @source [http://www.bls.gov/lau/rdscnp16.htm]()
"state_year"

#' @title State by month level US employment data
#'
#' @description Monthly series (seasonly adjusted) of employment status of the civilian noninstitutional population from BLS.
#'
#' @format A data frame with ten variables:
#' * `fips` = "FIPS code",
#' * `state` = "State or area",
#' * `year` = "Year",
#' * `month` = "Month",
#' * `pop` = "Civilian non-institutional population",
#' * `clf` = "Total numer of people in civilian labor force",
#' * `pc_clf` = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
#' * `emp` = "Total number of people employed",
#' * `pc_emp` = "Employment-population ratio (= employment / population; Age: 16 years and over)",
#' * `unem` = "Total number of people unemployed",
#' * `unem_rate` = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
#'
#' @source [http://www.bls.gov/lau/rdscnp16.htm]()
"state_month_sa"

#' @title State by month level US employment data
#'
#' @description Monthly series (not seasonly adjusted) of employment status of the civilian noninstitutional population from BLS.
#'
#' @format A data frame with ten variables:
#' * `fips` = "FIPS code",
#' * `state` = "State or area",
#' * `year` = "Year",
#' * `month` = "Month",
#' * `pop` = "Civilian non-institutional population",
#' * `clf` = "Total number of people in civilian labor force",
#' * `pc_clf` = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
#' * `emp` = "Total number of people employed",
#' * `pc_emp` = "Employment-population ratio (= employment / population; Age: 16 years and over)",
#' * `unem` = "Total number of people unemployed",
#' * `unem_rate` = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
#'
#' @source [http://www.bls.gov/lau/rdscnp16.htm]()
"state_month_nsa"

#' @title County-Year
#'
#' @description Annual series of US county employment data
#'
#' @format A data frame with ten variables:
#' * `laus_code` = "LAUS Code",
#' * `fips` = "County FIPS",
#' * `state_fips` = "State FIPS Code",
#' * `county_fips` = "County FIPS Code",
#' * `county` = "County Name",
#' * `state` = "State Abbreviation",
#' * `year` = "Year",
#' * `labor_force` = "Total numer of people in civilian labor force",
#' * `employed` = "Total number of people employed"
#' * `unemployed` = "Total number of people unemployed",
#' * `unem_rate` = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
#'
#' @source [https://www.bls.gov/lau/]()
"county_year"

#' @title County-Month
#'
#' @description Monthly series (not seasonly adjusted) of employment status of the civilian noninstitutional population from BLS.
#'
#'
#' @source [https://www.bls.gov/lau/]()
"county_month_nsa"


data-raw/prep_state_year.R
# Employment status of the civilian noninstitutional population, annual averages series

library(tidyverse)

# Downloade Statewide Annual Averages -------------------------------------

url = "http://www.bls.gov/lau/staadata.zip"
fil = "data-raw/state_year/staadata.zip"
downloader::download(url, fil)
unzip(fil, exdir = "data-raw/state_year")
list.files("data-raw/state_year")

# Load the data, add variable name, and variable labels -------------------

readxl::read_excel(path = "data-raw/state_year/staadata.xlsx",
                   col_name = c("fips", "state", "year", "pop", "clf", "pc_clf", "emp", "pc_emp", "unem", "unem_rate"),
                   skip = 8) %>%
    print() -> state_year

state_year %>% count(year) %>% print(n = 43)

labelled::var_label(state_year) = list(
        fips = "FIPS code",
        state = "State or area",
        year = "Year",
        pop = "Civilian non-institutional population",
        clf = "Total number of people in civilian labor force",
        pc_clf = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
        emp = "Total number of people employed",
        pc_emp = "Employment-population ratio (= employment / population; Age: 16 years and over)",
        unem = "Total number of people unemployed",
        unem_rate = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
)

# Drop LA and NYC ---------------------------------------------------------

state_year %>%
    anti_join(fips::state)

state_year %>%
    filter(fips != "037" & fips != "51000") %>%
    print() -> state_year

# Save it! ----------------------------------------------------------------

devtools::use_data(state_year, overwrite = TRUE)


data-raw/prep_county_year.R
# Labor force data by county, annual averages, 1990-2016

library(tidyverse)

# Download all Excel files ------------------------------------------------

urls = paste0("https://www.bls.gov/lau/laucnty", c(90:99, formatC(0:18, width=2, flag=0)), ".xlsx")
fils = paste0("data-raw/county_year/", basename(urls))
map2(urls, fils, download.file)

# Check out a few files ---------------------------------------------------

fils[1] %>%
    readxl::read_excel(col_names = FALSE, skip = 5) %>%
    select(-X__6) %>%         # Drop An empty column
    filter(X__2 != "NA") %>%  # Drop the last few rows
    rename(laus_code = X__1,
           state_fips = X__2,
           county_fips = X__3,
           county = X__4,
           year = X__5,
           labor_force = X__7,
           employed = X__8,
           unemployed = X__9,
           unemployment_rate = X__10) %>%
    separate(county, c("county", "state"), sep = ", ") %>% # Doesn't work for DC
    unite(fips, state_fips, county_fips, sep = "", remove = FALSE)

# Error message in row 317
readxl::read_excel("data-raw/county_year/laucnty90.xlsx", col_names = FALSE, skip = 5)[317,]

# Load all Excel files ----------------------------------------------------

tidy <- . %>%
    readxl::read_excel(col_names = FALSE, skip = 5) %>%
    select(-X__6) %>%         # Drop An empty column
    filter(X__2 != "NA") %>%  # Drop the last few rows
    rename(laus_code = X__1,
           state_fips = X__2,
           county_fips = X__3,
           county = X__4,
           year = X__5,
           labor_force = X__7,
           employed = X__8,
           unemployed = X__9,
           unemployment_rate = X__10) %>%
    separate(county, c("county", "state"), sep = ", ") %>% # Doesn't work for DC
    unite(fips, state_fips, county_fips, sep = "", remove = FALSE)

map(fils, tidy) %>% map(head) # Warnings are either about DC or missing values

map_df(fils, tidy) -> county_year
county_year

# Save it! ----------------------------------------------------------------

usethis::use_data(county_year, overwrite = TRUE)


data-raw/prep_county_month.R

# Labor force data by county, annual averages, 1990-01 - 2019-02

library(tidyverse)

# Meta --------------------------------------------------------------------

urls = c("https://download.bls.gov/pub/time.series/la/la.period",
    "https://download.bls.gov/pub/time.series/la/la.state_region_division",
    "https://download.bls.gov/pub/time.series/la/la.measure",
    "https://download.bls.gov/pub/time.series/la/la.footnote",
    "https://download.bls.gov/pub/time.series/la/la.area_type",
    "https://download.bls.gov/pub/time.series/la/la.area")

map(urls, ~ read_tsv(.x) %>% assign(basename(.x), ., envir = .GlobalEnv))

# County monthly laus -----------------------------------------------------

read_tsv("https://download.bls.gov/pub/time.series/la/la.data.64.County") %>%
    print() -> raw

# Breakdown series id -----------------------------------------------------
# See https://download.bls.gov/pub/time.series/la/la.txt

raw %>%
    mutate(area_code = substr(series_id, 4, 18)) %>%
    mutate(seasonal = substr(series_id, 3, 3)) %>%
    mutate(fips = substr(series_id, 6, 10)) %>%
    mutate(state_fips = substr(series_id, 6, 7)) %>%
    mutate(county_fips = substr(series_id, 8, 10)) %>%
    mutate(measure_code = substr(series_id, nchar(series_id) - 1, nchar(series_id))) %>%
    print(n = 20) -> raw2

# Add area_text -----------------------------------------------------------

la.area %>%
    select(area_code, area_text) %>%
    right_join(raw2) %>%
    mutate(area_text = if_else(area_text == "District of Columbia",
        paste0(area_text, ", DC"), area_text)) %>%
    separate(area_text, c("county", "state"), sep = ", ") %>%
    select(-series_id, -area_code) %>%
    select(fips, state_fips, county_fips, county, state, year, everything()) %>%
    print(n = 20) -> raw3

# Spread by measures ------------------------------------------------------

raw3 %>%
    left_join(la.measure) %>%
    select(-measure_code) %>%
    spread(measure_text, value) %>%
    print(n = 20) -> raw4

# Add period and footnote codes -------------------------------------------

raw4 %>%
    left_join(la.period) %>%
    mutate(month = substr(period, 2, 3)) %>%
    mutate(month = if_else(month == 13, NA_character_, month)) %>%
    print(n = 20) %>%
    left_join(la.footnote, by = c("footnote_codes" = "footnote_code")) %>%
    select(-footnote_codes) %>%
    select(fips, state_fips, county_fips,
           county, state, year, month,
           employment, `labor force`, unemployment, `unemployment rate`,
           seasonal, everything()) %>%
    print(n = 20) -> raw5

# Don’t include annual average --------------------------------------------
# And don't include 2019-03
raw5 %>% filter(!is.na(month)) -> county_month_nsa

county_month_nsa %>% count(year, month) %>% print(n = 351)
county_month_nsa %>% filter(!(year == 2019 & month == "03")) -> county_month_nsa

usethis::use_data(county_month_nsa, overwrite = TRUE)


data-raw/prep_state_month_sa.R
# Employment status of the civilian noninstitutional population, seasonaly adjusted monthly series
# 1976-01 - 2019-03
# For more information, see https://www.bls.gov/web/laus.supp.toc.htm

library(tidyverse)

# Downloade Statewide Annual Averages -------------------------------------

url = "https://www.bls.gov/web/laus/ststdsadata.zip"
fil = "data-raw/state_month/ststdsadata.zip"
downloader::download(url, fil)
unzip(fil, exdir = "data-raw/state_month")
list.files("data-raw/state_month")

# Load the data, add variable name, and variable labels -------------------

readxl::read_excel(path = "data-raw/state_month/ststdsadata.xlsx",
                   col_name = c("fips", "state", "year", "month", "pop", "clf", "pc_clf", "emp", "pc_emp", "unem", "unem_rate"),
                   skip = 8) %>%
    print() -> state_month_sa

state_month_sa %>% count(year, month) %>% print(n = 519)

labelled::var_label(state_month_sa) = list(
        fips = "FIPS code",
        state = "State or area",
        year = "Year",
        month = "Month",
        pop = "Civilian non-institutional population",
        clf = "Total number of people in civilian labor force",
        pc_clf = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
        emp = "Total number of people employed",
        pc_emp = "Employment-population ratio (= employment / population; Age: 16 years and over)",
        unem = "Total number of people unemployed",
        unem_rate = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
)

# Drop LA and NYC ---------------------------------------------------------

state_month_sa %>%
    anti_join(fips::state)

state_month_sa %>%
    filter(fips != "037" & fips != "51000") %>%
    print() -> state_month_sa

# Save it! ----------------------------------------------------------------

usethis::use_data(state_month_sa, overwrite = TRUE)


data-raw/prep_state_month_nsa.R
# Employment status of the civilian noninstitutional population, not seasonaly adjusted monthly series
# 1976-01 - 2019-03
# For more information, see https://www.bls.gov/web/laus.supp.toc.htm

library(tidyverse)

# Downloade Statewide Annual Averages -------------------------------------

url = "https://www.bls.gov/web/laus/ststdnsadata.zip"
fil = "data-raw/state_month/ststdnsadata.zip"
downloader::download(url, fil)
unzip(fil, exdir = "data-raw/state_month")
list.files("data-raw/state_month")

# Load the data, add variable name, and variable labels -------------------

readxl::read_excel(path = "data-raw/state_month/ststdnsadata.xlsx",
                   col_name = c("fips", "state", "year", "month", "pop", "clf", "pc_clf", "emp", "pc_emp", "unem", "unem_rate"),
                   skip = 8) %>%
    print() -> state_month_nsa

state_month_nsa %>% count(year, month) %>% print(n = 519)

labelled::var_label(state_month_nsa) = list(
        fips = "FIPS code",
        state = "State or area",
        year = "Year",
        month = "Month",
        pop = "Civilian non-institutional population",
        clf = "Total number of people in civilian labor force",
        pc_clf = "Labor force participation rate (= labor force / population; Age: 16 years and over)",
        emp = "Total number of people employed",
        pc_emp = "Employment-population ratio (= employment / population; Age: 16 years and over)",
        unem = "Total number of people unemployed",
        unem_rate = "Unemployment rate (= unemployment / labor force; Age: 16 years and over)"
)

# Drop LA and NYC ---------------------------------------------------------

state_month_nsa %>%
    anti_join(fips::state)

state_month_nsa %>%
    filter(fips != "037" & fips != "51000") %>%
    print() -> state_month_nsa

# Save it! ----------------------------------------------------------------

usethis::use_data(state_month_nsa, overwrite = TRUE)
